x/stake/keeper/val_state_change.go

```diff
@@ -0,0 +1,262 @@
+			panic("should never retrieve a jailed validator from the power store")
+		}
+
+		// if we get to a zero-power validator (which we don't bond),
+		// there are no more possible bonded validators
+		// note: we must check the ABCI power, since we round before sending to Tendermint
+		if validator.Tokens.RoundInt64() == int64(0) {
+			break
+		}
+
+		// apply the appropriate state change if necessary
+		switch validator.Status {
+		case sdk.Unbonded:
+			validator = k.unbondedToBonded(ctx, validator)
+		case sdk.Unbonding:
+			validator = k.unbondingToBonded(ctx, validator)
+		case sdk.Bonded:
+			// no state change
+		}
+
+		// fetch the old power bytes
+		var opbytes [sdk.AddrLen]byte
+		copy(opbytes[:], operator[:])
+		oldPowerBytes, ok := last[opbytes]
+
+		// calculate the new power bytes
+		newPowerBytes := validator.ABCIValidatorPowerBytes(k.cdc)
+
+		// update the validator set if power has changed
+		if !ok || !bytes.Equal(oldPowerBytes, newPowerBytes) {
+			updates = append(updates, validator.ABCIValidatorUpdate())
+		}
+
+		// validator still in the validator set, so delete from the copy
+		delete(last, opbytes)
+
+		// set the bonded validator index
+		store.Set(GetBondedValidatorIndexKey(operator), newPowerBytes)
+
+		// keep count
+		count++
+
+	}
+
+	// sort the no-longer-bonded validators
+	noLongerBonded := k.sortNoLongerBonded(last)
+
+	// iterate through the sorted no-longer-bonded validators
+	for _, operator := range noLongerBonded {
+
+		// fetch the validator
+		validator := k.mustGetValidator(ctx, sdk.ValAddress(operator))
+
+		// bonded to unbonding
+		k.bondedToUnbonding(ctx, validator)
+
+		// remove validator if it has no more tokens
+		if validator.Tokens.IsZero() {
+			k.RemoveValidator(ctx, validator.OperatorAddr)
+		}
+
+		// delete from the bonded validator index
+		store.Delete(GetBondedValidatorIndexKey(operator))
+
+		// update the validator set
+		updates = append(updates, validator.ABCIValidatorUpdateZero())
+
+	}
+
+	return updates
+}
+
+// Validator state transitions
+
+func (k Keeper) bondedToUnbonding(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Bonded {
+		panic(fmt.Sprintf("bad state transition bondedToUnbonding, validator: %v\n", validator))
+	}
+	return k.beginUnbondingValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonding {
+		panic(fmt.Sprintf("bad state transition unbondingToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondedToBonded(ctx sdk.Context, validator types.Validator) types.Validator {
+	if validator.Status != sdk.Unbonded {
+		panic(fmt.Sprintf("bad state transition unbondedToBonded, validator: %v\n", validator))
+	}
+	return k.bondValidator(ctx, validator)
+}
+
+func (k Keeper) unbondingToUnbonded(ctx sdk.Context, validator types.Validator) types.Validator {
```

Comment:
this is currently unused however I believe this will need to be used in the validator queue (CC @sunnya97 ) so let's leave it

Reply:
Yup, left for that reason.